src/VisualStudio/Core/Def/ErrorReporting/VisualStudioInfoBar.cs

```diff
@@ -4,119 +4,174 @@
-    /// same message in the future.  This is important for user clarity as it's possible for 
-    /// a feature to fail for some reason, then work fine for a while, then fail again.  We want
-    /// the second failure message to be reported to ensure the user is not confused.
+    /// Keep track of the messages that are currently being shown to the user.  If we would show the same message again,
+    /// block that from happening so we don't spam the user with the same message.  When the info bar item is dismissed
+    /// though, we then may show the same message in the future.  This is important for user clarity as it's possible
+    /// for a feature to fail for some reason, then work fine for a while, then fail again.  We want the second failure
+    /// message to be reported to ensure the user is not confused.
     /// 
-    /// Accessed on UI thread.
+    /// Accessed on UI thread only.
     /// </summary>
     private readonly HashSet<string> _currentlyShowingMessages = [];
 
-    public VisualStudioInfoBar(
-        IThreadingContext threadingContext,
-        SVsServiceProvider serviceProvider,
-        IAsynchronousOperationListenerProvider listenerProvider)
-    {
-        _threadingContext = threadingContext;
-        _serviceProvider = serviceProvider;
-        _listener = listenerProvider.GetListener(FeatureAttribute.InfoBar);
-    }
+    public void ShowInfoBarMessageFromAnyThread(string message, params InfoBarUI[] items)
+        => ShowInfoBarMessageFromAnyThread(message, isCloseButtonVisible: true, KnownMonikers.StatusInformation, items);
 
-    public void ShowInfoBar(string message, params InfoBarUI[] items)
+    public void ShowInfoBarMessageFromAnyThread(
+        string message,
+        bool isCloseButtonVisible,
+        ImageMoniker imageMoniker,
+        params InfoBarUI[] items)
     {
         // We can be called from any thread since errors can occur anywhere, however we can only construct and InfoBar from the UI thread.
         _threadingContext.JoinableTaskFactory.RunAsync(async () =>
         {
-            using var _ = _listener.BeginAsyncOperation(nameof(ShowInfoBar));
+            using var _ = _listener.BeginAsyncOperation(nameof(ShowInfoBarMessageFromAnyThread));
 
             await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(_threadingContext.DisposalToken);
+            ShowInfoBarMessageFromUIThread(message, isCloseButtonVisible, imageMoniker, items);
+        });
+    }
+
+    public InfoBarMessage? ShowInfoBarMessageFromUIThread(
+        string message,
+        bool isCloseButtonVisible,
+        ImageMoniker imageMoniker,
+        params InfoBarUI[] items)
+    {
+        _threadingContext.ThrowIfNotOnUIThread();
+
+        // If we're already shown this same message to the user, then do not bother showing it
+        // to them again.  It will just be noisy.
+        if (GetInfoBarHostObject() is not IVsInfoBarHost infoBarHost ||
```

Comment:
Tweak: Existing code always put the info bar in the global location.  Now it can put it at the global location, or on a specific window frame.  this is used by SG code.